### PR TITLE
Update bytes to v0.6

### PR DESCRIPTION
**Pull request by olix0r**
Depends on:

- [ ] hyperium/h2#497
- [ ] hyperium/http-body#29
- [ ] hyperium/http#440

### CODE REVIEW

**Comment by Urhengulas** [comment]
> This doesn't currently compile due to the tokio::ReadBuf API.

@olix0r: Thank you for taking ownership for this :) Can you please elaborate what's holding the PR back?

**Comment by olix0r** [comment]
@Urhengulas Sorry, that comment is outdated (addressed in 34e60f7). It looks like there are some remaining changes needed in examples that I can try to fix. I won't be offended if someone else takes this branch and gets it over the line, though ;)

**Comment by olix0r** [comment]
Also, this is blocked on a few other PRs:

* https://github.com/hyperium/h2/pull/498 -- not currently passing CI
* https://github.com/hyperium/http-body/pull/30
* https://github.com/hyperium/http/pull/441

**Comment by paolobarbolini** [comment]
My h2 PR does pass CI https://github.com/hyperium/h2/pull/497

**Comment by olix0r** [comment]
@paolobarbolini whoops! didn't mean to double up on those PRs, I just didn't look before branching. Closed my open PRs in favor of yours.

**Comment by paolobarbolini** [comment]
> It looks like there are some remaining changes needed in examples that I can try to fix

It probably just needs `bytes::buf::BufExt` to be changed into `bytes::Buf`

**Comment by olix0r** [comment]
OK, build passes. gitcop won't tell me what it doesn't like(?!), so i'm going to rebase and force push.

File: src/proto/h1/io.rs [hunk 1]
```diff
@@ -188,7 +189,9 @@ where
         if self.read_buf_remaining_mut() < next {
             self.read_buf.reserve(next);
         }
-        let mut buf = ReadBuf::uninit(&mut self.read_buf.bytes_mut()[..]);
+        let dst = self.read_buf.bytes_mut();
+        let dst = unsafe { &mut *(dst as *mut _ as *mut [MaybeUninit<u8>]) };
```
Comment (Urhengulas) [inline src/proto/h1/io.rs:193]: That line looks threatening..

Can you please add a comment saying what it is doing?

**Comment by seanmonstar** [comment]
Don't worry about gitcop (it's not "required"), it's just a reminder to me to fix the commit message format.

File: src/proto/h1/io.rs (hunk 1 above)
Comment (olix0r) [reply src/proto/h1/io.rs:193]: I don't actually know what it's doing ;)

I borrowed this from https://github.com/tokio-rs/tokio/blob/d78655337a68bded305782a8a8b4ac7be42aa6a7/tokio/src/io/util/read_buf.rs#L53-L55, which was the only updated use of `ReadBuf::uninit` I could find.

I would appreciate feedback who understand these APIs better than I do.

Comment (Urhengulas) [reply src/proto/h1/io.rs:193]: @seanmonstar ?

Comment (Urhengulas) [reply src/proto/h1/io.rs:193]: But let's see who introduced it 🙂

Comment (Urhengulas) [reply src/proto/h1/io.rs:193]: https://github.com/tokio-rs/tokio/discussions/3121

Comment (olix0r) [reply src/proto/h1/io.rs:195]: Thanks, added a comment summarizing that discussion.

File: src/proto/h1/io.rs [hunk 2]
```diff
@@ -188,7 +189,11 @@ where
         if self.read_buf_remaining_mut() < next {
             self.read_buf.reserve(next);
         }
-        let mut buf = ReadBuf::uninit(&mut self.read_buf.bytes_mut()[..]);
+        let dst = self.read_buf.bytes_mut();
+        // Safety: `dst` may include uninitialized memory, but `ReadBuf` will
+        // never uninitialize memory that has already been initialized.
+        let dst = unsafe { &mut *(dst as *mut _ as *mut [MaybeUninit<u8>]) };
+        let mut buf = ReadBuf::uninit(dst);
         match Pin::new(&mut self.io).poll_read(cx, &mut buf) {
```
Comment (hawkw) [inline src/proto/h1/io.rs:197]: An alternative to all of this (which avoids the scary-looking pointer cast that @Urhengulas was concerned about) is to just use `tokio_util::io::poll_read_buf` from the latest `tokio-util` here:
```suggestion
        match tokio_util::io::poll_read_buf(Pin::new(&mut self.io), cx, &mut self.read_buf) {
```

That would mean adding a `tokio-util` dependency, but it's already a transitive dependency via `h2`.

File: src/proto/h1/io.rs (hunk 1 above)
Comment (Urhengulas) [reply src/proto/h1/io.rs:195]: Perfect 😄

File: src/proto/h1/io.rs (hunk 2 above)
Comment (Urhengulas) [reply src/proto/h1/io.rs:197]: That seems like a good solution!

And this is a useful function which can't be found in the docs... (code: https://github.com/tokio-rs/tokio/blob/master/tokio-util/src/lib.rs#L67-L134).

**Comment by hawkw** [comment]
I believe #2339 closes this.

**Pull request closed without merging**